src/opensynth/models/faraday/new_gmm/new_gmm_model.py

```diff
@@ -281,73 +284,104 @@ def training_step(self, batch) -> None:
         # Run e-step
         log_prob, log_resp = self.gmm_module.e_step(encoded_batch)
         # Run m-step
-        precision_cholesky, weights, means, covar = self.gmm_module.m_step(
-            encoded_batch, log_resp
+        precision_cholesky, weights, means, covariances = (
+            self.gmm_module.m_step(encoded_batch, log_resp)
         )
         # Update model params. This only updates the params
         # on the current device
         self.gmm_module.update_params(
             weights=weights,
             means=means,
             precision_cholesky=precision_cholesky,
-            covariances=covar,
-            log_prob=log_prob,
+            covariances=covariances,
+            nll=torch.neg(log_prob),
         )
 
+        self.log(
```

Comment:
I'm not 100% sure if this should be the right way to log NLL. I'd have think we want to:

- Sync parameters across devices, so all devices now have the same weight
- Then only you calculate the NLL and check convergence

What you're doing here is calculate log prob across devices, then averaging the log prob. Don't think this should be the right way?

But perhaps it's fine because this step technically only happens _after_ syncing on batch/ syncing on epoch - so theoretically when performing e_step, log_prob should be similar across all devices anyway? The only difference across devices will just be a difference in the batches, in which case averaging should be fine.

Worth clarifying this as comments in doc, if what I think is happening is really happening!

Reply:
To add clarity to this, I have moved the logging to the `on_train_epoch_end` 
 and logged `nll_reduced` after metrics have been updated and computed.